packages/next/src/lib/metadata/resolvers/resolve-opengraph.ts

```diff
@@ -58,25 +59,41 @@ function resolveAndValidateImage(
   // detect in the build environment if the deployment is a Vercel deployment or not.
   //
   // x-ref: https://vercel.com/docs/projects/environment-variables/system-environment-variables#system-environment-variables
-  const isNonVercelDeployment = !process.env.VERCEL
-  // When not deployed to Vercel, we aren't able to determine a fallback value for `metadataBase`
-  // since it's derived from injected environment variables. For self-hosted setups, we want to warn
-  // about this since the only fallback we'll be able to generate is `localhost`.
-  if (isNonVercelDeployment) {
-    if (
-      typeof inputUrl === 'string' &&
-      !isFullStringUrl(inputUrl) &&
-      !metadataBase
-    ) {
+  const isUsingVercelSystemEnvironmentVariables = Boolean(process.env.VERCEL)
+
+  const isRelativeUrl =
+    typeof inputUrl === 'string' && !isFullStringUrl(inputUrl)
+
+  // We no explicit metadataBase is specified by the user, we'll override it with the fallback metadata
+  // under the following conditions:
+  // - The user didn't provide a metadataBase and the provided URL is relative (ie ./og-image).
+  // - The image is statically generated by Next.js (such as the special `opengraph-image` route)
+  // In both cases, we want to ensure that across all environments, the ogImage is a fully qualified URL.
+  // In the `opengraph-image` route case, since the user isn't explicitly passing a relative path, this ensures
+  // the ogImage will be properly discovered across different environments without the user needing to
+  // have a bunch of `process.env` checks when defining their `metadataBase`.
+  if (isRelativeUrl && (!metadataBase || isStaticMetadataRouteFile)) {
+    // When not using Vercel environment variables for URL injection, we aren't able to determine
+    // a fallback value for `metadataBase`. For self-hosted setups, we want to warn
+    // about this since the only fallback we'll be able to generate is `localhost`.
+    // In development, we'll only warn for relative metadata that isn't part of the static
+    // metadata conventions (eg `opengraph-image`), as otherwise it's currently very noisy
+    // for common cases. Eventually we should remove this warning all together in favor of
+    // devtools.
+    const shouldWarn =
+      !isUsingVercelSystemEnvironmentVariables &&
+      (process.env.NODE_ENV === 'production'
```

Comment:
can be shorter
`(process.env.NODE_ENV === 'production' || !isStaticMetadataRouteFile)`